### DISCUSSION TITLE
copytoclipboard Quasar utils example

### DISCUSSION THREAD

**Discussion by Xtreemrus** [zauberzeug/nicegui#1679] [Ideas / Feature Requests]
Good day,
Unfortunately, I couldn't figure out how to add Quasar clipboard support using NiceGUI.
https://quasar.dev/quasar-utils/other-utils#copytoclipboard

It would be great to add an example to the documentation, because the copy button click is almost on every website and is quite popular.

Now on the forum you can find example like code:
```py
await ui.run_javascript(f"""navigator.clipboard.writeText("{answer_text}")""", respond=False)
```

But from what I understood, despite the seeming simplicity, making the copyToClipboard button work on PC, Mac, iOS, and Android on all browsers may not be that simple.
Not to mention that in theory, you can set different clipboards to store both formatted text and plain text.

That's why the idea came that it's better to entrust this to Quasar, rather than trying to track which new version of iOS will break some workaround or chrome will revise the clipboard standard.

Here's a funny option that GPT suggested for copying formatted text.
For a Python developer, such constructions just to copy text look quite unusual :)

```py
    await ui.run_javascript(f"""
        var div = document.createElement('div');
        div.innerHTML = `{str(answer_text_formatted)}`;
        div.contentEditable = true;
        div.style.position = 'fixed';
        div.style.left = '-9999px';
        document.body.appendChild(div);
        var range = document.createRange();
        range.selectNodeContents(div);
        var selection = window.getSelection();
        selection.removeAllRanges();
        selection.addRange(range);
        document.execCommand('copy');
        document.body.removeChild(div);
    """, respond=False)
```

**Comment by falkoschindler** [comment]
Why do you think `navigator.clipboard.writeText` wouldn't work cross-platform?
Browser compatibility looks good: https://developer.mozilla.org/en-US/docs/Web/API/Navigator/clipboard#browser_compatibility

**Reply by Xtreemrus** [reply]
My knowledge in frontend is quite modest, so when the code didn't work, I started looking for how JavaScript works with clipboard and saw a lot of different workarounds and nuances, including those on iOS.

https://wolfgangrittner.dev/how-to-use-clipboard-api-in-safari/
https://developer.apple.com/forums/thread/691873

And when I looked at Quasar and their example worked on iOS and on all browsers just by clicking on the link.

Almost half of Google's answers to the query "ios navigator.clipboard.writeText" lead to various articles on how to make it work or that it doesn't work.

That's why I thought that trying to overcome this in Python, and then keeping it up to date is quite difficult.

Part of the problem might also be that I'm not testing on https (it's local http://192.168.0.190:8080), as far as I understand this also affects https://stackoverflow.com/questions/51805395/navigator-clipboard-is-undefined . Overall, I'm very surprised that clipboard writing is such a deep topic in 2023 for JavaScript developers and browser manufacturers.

**Comment by coolynx** [comment]
It works in browser, but I could not get it work in Native mode using PyWebView, because of permission.

```
[Error] Unhandled Promise Rejection: 
NotAllowedError: The request is not allowed by the user agent 
or the platform in the current context, possibly because the user denied permission.
	promiseEmptyOnRejected
	promiseReactionJob
```
It needs some kind of setting that maybe will be available in future PyWebView v5.0 version.

https://github.com/r0x0r/pywebview/issues/1252#issuecomment-1805307798

```py
webkit.settings.javascript_can_access_clipboard = True
```
